File: 3a_Run-time-Polymorphism_ADT/include/duck.r
#ifndef DUCK_R
#define DUCK_R

#define MAX_CHARS_NAME 10
#define MAX_CHARS_NAME_WITH_NUL (MAX_CHARS_NAME+1)

typedef struct Duck_Interface_Struct const * Duck_Interface;

typedef struct Duck_t
{
    Duck_Interface vtable;
    char name[MAX_CHARS_NAME_WITH_NUL];
} Duck_t;

typedef struct Duck_Interface_Struct
{
    void (*show)( Duck_t * thisDuck );  // Optional for derived classes; set to "0" or "NULL" to use the default "Duck" implementation
    void (*deinit)( Duck_t * thisDuck ); // Optional for derived classes; set to "0" or "NULL" if no deinitialization of the derived class is required
    void (*destroy)( Duck_t * thisDuck );
} Duck_Interface_Struct;

#endif // DUCK_R

File: 3_Run-time-Polymorphism/d_Making-duck-an-abstract-class/source/duck.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "assert.h"
#include "duck.h"
#include "duck_private.h"

void
duckQuack( p_Duck_t thisDuck )
{
    ASSERT(thisDuck && thisDuck->name);
    
    printf("\t%s: Quack!!\n", thisDuck->name);
}

void
duckShow( p_Duck_t thisDuck )
{
    ASSERT(thisDuck && thisDuck->name && thisDuck->vtable && thisDuck->vtable->show);

    thisDuck->vtable->show(thisDuck);
}

void
duckDestroy( p_Duck_t thisDuck )
{
    ASSERT(thisDuck && thisDuck->name && thisDuck->vtable);
    ASSERT(thisDuck->vtable->deinit && thisDuck->vtable->destroy);

    printf("\tDestroying Duck object with name: %s\n", thisDuck->name);
    
    thisDuck->vtable->deinit(thisDuck);
    
    // Duck deinitialization
    memset(thisDuck->name, 0, MAX_CHARS_NAME);

    thisDuck->vtable->destroy(thisDuck);
}